File: Medium/2958. Length of Longest Subarray With at Most K Frequency.java
class Modified {
    public int maxSubarrayLength(int[] nums, int k) {
        var map = new HashMap<Integer, ArrayDeque<Integer>>();
        int ans = 0;
        for (int l = 0, r = 0; r < nums.length; r++) {
            map.computeIfAbsent(nums[r], key -> new ArrayDeque<>()).offer(r);
            if (map.get(nums[r]).size() > k) {
                if (map.get(nums[r]).peekFirst() >= l) {
                    l = map.get(nums[r]).peekFirst() + 1;
                }
                map.get(nums[r]).poll();
            }
            ans = Math.max(ans, r - l + 1);
        }
        return ans;
    }
}

class Naive {
    public int maxSubarrayLength(int[] nums, int k) {
        var map = new HashMap<Integer, Integer>();
        int res = 0;
        for (int l = 0, r = 0; r < nums.length; r++) {
            map.put(nums[r], map.getOrDefault(nums[r], 0) + 1);
            while (map.get(nums[r]) > k) {
                map.put(nums[l], map.get(nums[l]) - 1);
                l++;
            }
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}

File: Medium/2958. Length of Longest Subarray With at Most K Frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        freq = defaultdict(int)
        res = l = 0
        for r in range(len(nums)):
            freq[nums[r]] += 1
            while freq[nums[r]] > k:
                freq[nums[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res


class Solution1:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        mapp = defaultdict(deque)
        res = l = 0
        for r in range(n):
            mapp[nums[r]].append(r)
            curr = mapp[nums[r]]
            if len(curr) > k:
                l = max(curr.popleft() + 1, l)
            res = max(r - l + 1, res)
        return res
